src/Admin/HostedServices/AzureQueueMailHostedService.cs

```diff
@@ -70,17 +69,19 @@ private async Task ExecuteAsync(CancellationToken cancellationToken)
                 {
                     try
                     {
-                        var token = JToken.Parse(message.DecodeMessageText());
-                        if (token is JArray)
+                        using var document = JsonHelpers.Parse(message.DecodeMessageText());
+                        var root = document.RootElement;
+
+                        if (root.ValueKind == JsonValueKind.Array)
                         {
-                            foreach (var mailQueueMessage in token.ToObject<List<MailQueueMessage>>())
+                            foreach (var mailQueueMessage in root.ToObject<List<MailQueueMessage>>())
                             {
                                 await _mailService.SendEnqueuedMailMessageAsync(mailQueueMessage);
                             }
                         }
-                        else if (token is JObject)
+                        else if (root.ValueKind == JsonValueKind.Object)
```

Comment:
I think that makes sense... Is it an issue with self-hosted at all since they don't use the queues? I like Justin's idea of phasing the enqueue logic first using Arrays, then next release update the dequeue logic to match.

Reply:
Self-hosted won't be effected since they are table-backed for events and use a blocking mail service instead of the Azure queue for bulk emails.